tensilelite/Tensile/KernelWriterAssembly.py

```diff
@@ -4125,6 +4125,89 @@ def lraSwapAddressesForDTLPad(self, kernel, tP):
                          comment="xor both lds buffer offsets to enable swapping"))
     return module
 
+  def checkAlignmentForDTLTailLoop(self, kernel, tPA, tPB):
+    # For 16bit/8bit floats we need to check for alignment and modify the
+    # buffer loads when using wider (dword,dwordx4) loads to workaround OOB behaviours
+    #
+    # For 16b:
+    # if k even: nothing needs to be done
+    # else k odd:
+    #   if addr 4B aligned: add 2B to size record in buf desc
+    #   else: shift gro left by 2B and shift lro right by 2B
+    #
+    # For 8b: TBD.
+    #
+    module = Module("Check Alignment For DTL Tail Loop")
+
+    if not kernel["DirectToLdsA"] and not kernel["DirectToLdsB"]:
+      return Module()
+
+    # Skip this for 8bit types for now.
+    if kernel["ProblemType"]["DataTypeA"].isAnyFloat8() or kernel["ProblemType"]["DataTypeB"].isAnyFloat8():
+      return Module()
+
+    module.addComment0("Check if global reads need to be modified for DTL Tail Loop")
+
+    tmpSgpr = self.sgprPool.checkOut(1)
+    labelSkip = Label("SkipDTLTailLoopGRChanges", "" )
+    goodLMultiple = 2
+
+    module.add(SAndB32(dst=sgpr(tmpSgpr), src0=sgpr("SizeL"), src1=hex(goodLMultiple - 1), comment="Check if K is multiple of %u"%goodLMultiple))
+    module.add(SCmpEQU32(src0=sgpr(tmpSgpr), src1=hex(0), comment="" ))
+    self.sgprPool.checkIn(tmpSgpr)
+
+    module.add(SCBranchSCC1(labelName=labelSkip.getLabelName(), comment="Skip global read changes" ))
+
+    def offsetCalcDTLTailloop(kernel, tP):
+      tc = tP["tensorChar"]
+
+      if not kernel["DirectToLds%s"%tc]:
+        return Module()
+
+      module = Module("DTL OOB checks for %s"%tc)
+      dataType = kernel["ProblemType"]["DataType%s"%tc]
+
+      if dataType.isHalf() or dataType.BFloat16():
+        tmpSgpr = self.sgprPool.checkOut(1)
+        labelEnd = Label("EndDTLTailLoopGRChanges%s"%tc, "" )
+        labelOffsetMod = Label("DTLOffsetMod%s"%tc, "" )
+
+        padValue = 2
+
+        module.add(SAndB32(dst=sgpr(tmpSgpr), src0=sgpr("Srd%s+%u"%(tc,0)), src1=hex(0x3), comment="Check if addr is 4B aligned"))
+        module.add(SCmpEQU32(src0=sgpr(tmpSgpr), src1=hex(0), comment="" )) # check if 4B aligned
+        module.add(SCBranchSCC0(labelName=labelOffsetMod.getLabelName(), comment="" ))
+
+        # Case where address is 4B aligned
+        module.add(SCmpEQU32(src0=sgpr("Srd%s+%u"%(tc,2)), src1="BufferLimit", comment="" )) # check if 4B aligned
```

Comment:
We should not change Srd2.
That means we allow out of boundary access.

Reply:
Sorry. The previous comment is not true.

Reply:
I do not think we need to check BufferLimit in TailLoop.
Srd2 should be within BufferLimit in TailLoop (since remaining K is smaller than DepthU)

Reply:
Hmm...
K < DepthU is related to TLU=True case.
TLU=False case, Srd2 should be always within BufferLimit.

Reply:
> That means we allow out of boundary access.

Yes, the padding is to allow potential OOB read by 2 bytes. For dword{x4} loads the OOB check work at a 32bit granularity. Given some offset a thread will read 32bits starting from that offset and if OOB occurs in between that 32bit region, the load will return zero for the entire 32bit even if part of the read is in bounds. For K odd, this will lead to some values towards the end of the matrix to not be read correctly.